cmd/dep/status.go

```diff
@@ -335,6 +388,128 @@ func (cmd *statusCommand) validateFlags() error {
 	return nil
 }
 
+// OldStatus contains information about all the out of date packages in a project.
+type OldStatus struct {
+	ProjectRoot string
+	Constraint  gps.Constraint
+	Revision    gps.Revision
+	Latest      gps.Version
+}
+
+type rawOldStatus struct {
+	ProjectRoot, Constraint, Revision, Latest string
+}
+
+func (os OldStatus) getConsolidatedConstraint() string {
+	var constraint string
+	if os.Constraint != nil {
+		if v, ok := os.Constraint.(gps.Version); ok {
+			constraint = formatVersion(v)
+		} else {
+			constraint = os.Constraint.String()
+		}
+	}
+	return constraint
+}
+
+func (os OldStatus) getConsolidatedLatest(revSize uint8) string {
+	latest := ""
+	if os.Latest != nil {
+		switch revSize {
+		case shortRev:
+			latest = formatVersion(os.Latest)
+		case longRev:
+			latest = os.Latest.String()
+		}
+	}
+	return latest
+}
+
+func (os OldStatus) marshalJSON() *rawOldStatus {
+	return &rawOldStatus{
+		ProjectRoot: os.ProjectRoot,
+		Constraint:  os.getConsolidatedConstraint(),
+		Revision:    string(os.Revision),
+		Latest:      os.getConsolidatedLatest(longRev),
+	}
+}
+
+func (cmd *statusCommand) runOld(ctx *dep.Ctx, out oldOutputter, p *dep.Project, sm gps.SourceManager) error {
+	// While the network churns on ListVersions() requests, statically analyze
+	// code from the current project.
+	ptree, err := p.ParseRootPackageTree()
+	if err != nil {
+		return err
+	}
+
+	// Set up a solver in order to check the InputHash.
+	params := gps.SolveParameters{
+		ProjectAnalyzer: dep.Analyzer{},
+		RootDir:         p.AbsRoot,
+		RootPackageTree: ptree,
+		Manifest:        p.Manifest,
+		// Locks aren't a part of the input hash check, so we can omit it.
+	}
+
+	// Check update for all the projects
+	params.ChangeAll = true
+
+	solver, err := gps.Prepare(params, sm)
+	if err != nil {
+		return errors.Wrap(err, "fastpath solver prepare")
+	}
+
+	solution, err := solver.Solve(context.TODO())
+	if err != nil {
+		return errors.Wrap(err, "runOld")
+	}
+
+	var oldStatuses []OldStatus
+	solutionProjects := solution.Projects()
+
+	for _, proj := range p.Lock.Projects() {
+		for i := range solutionProjects {
```

Comment:
any reason for using an index number here and not the object directly?
`for _, solnProj := solutionProjects {`

I see `i` is used below only to get the object at `i`th position.

Reply:
https://github.com/golang/dep/pull/1402/commits/7a9180f5566f410af306f2854a32bd99315b058e

Reply:
@azbshiri Right, part of an optimization. Makes sense. Thanks.

Reply:
@zkry check if you can adopt that optimization and make it fast. Can be done separately once the logic is corrected.

Reply:
Ok, I can do that. The reason I changed it to a nested loop was because I found the order for the lock and solution projects to be different. I'll have to check where/why they are not in the same order. Whatever the case, sorting the list that is out of order and then comparing them with one iteration would be faster.

Reply:
So it turns out that solutionProjects is not sorted. I can sort it like it's done in runStatusAll (https://github.com/golang/dep/pull/1553/files#diff-cf328f431a592d0b360d509c6c5862c7R640). 

On a side note, since locked projects is being sorted for good measures here, do you think I should also sort it runOld?

Reply:
Benefit of sorting would be a consistent output, which is required for the tests to pass always. So yes, sort them.

Reply:
@zkry can you change the loop to use the array element and not the index as shown in the first comment in this thread? I don't see the need of index, as discussed above, it came from some old code which we don't need any more.
Also, sorry for so much delay. I'll be able to get this merged as soon as you make the change. Or if you are busy, I can make the changes myself :)

Reply:
No problem 🙂. I'll get this one in right away.